agenthub/codeact_agent/action_parser.py

```diff
@@ -157,10 +157,11 @@ def parse(self, action_str: str) -> Action:
         assert (
             self.agent_delegate is not None
         ), 'self.agent_delegate should not be None when parse is called'
-        thought = action_str.replace(self.agent_delegate.group(0), '').strip()
+        # thought = action_str.replace(self.agent_delegate.group(0), '').strip()
         browse_actions = self.agent_delegate.group(1).strip()
-        task = f'{thought}. I should start with: {browse_actions}'
```

Comment:
Was this intentionally deleted?

Reply:
Yes! This is intentional to make sure the "instruction" can be delegated correctly.

Reply:
The problem is it will lose the `thought`, which often contains the majority of the information needed.

Reply:
For example, if you look at `tests/integration/mock/CodeActAgent/test_browse_internet/prompt_002.log`, you can see the thought is `Certainly! I'll browse localhost:8000 and retrieve the ultimate answer to life for you.` and the `browse_actions` is `Get the content on "http://localhost:8000`, while the original task is `Browse localhost:8000, and tell me the ultimate answer to life. Do not ask me for confirmation at any point.`

Reply:
Yeah that's indeed a concern... But given that agent delegation is not enabled by default - is it a big issue now? I imagine we need to rework the delegation part with a lot of eval (to enable it by default) once we finish the Runtime refractor..

Reply:
Yep, the idea here is that we need to 1:1 copy over the instruction to `BrowisngAgent`, so we don't need to re-run all the browsing eval again to say that `CodeAct + delegation to BrowsingAgent` would perform the same as `BrowsingAgent` alone.. So including "thoughts" will change the input to `BrowsingAgent` (compared to it being evaluated alone), which deviate a bit from our purpose.

Reply:
> But given that agent delegation is not enabled by default

It is enabled by default

> the idea here is that we need to 1:1 copy over the instruction to BrowisngAgent

I don't think the current code does that either. Here `browse_actions` won't be the same as original user input.

Reply:
> I don't think the current code does that either. Here browse_actions won't be the same as original user input.

We are trying to test whether the agent can copy over the instructions (as instructed) here:
https://github.com/OpenDevin/OpenDevin/pull/2948/files#diff-031ec786aadff892d31ee4af114fa57ed566ca3b08e8a87cc9a7d62a681df7e8R60-R65

> It is enabled by default (there's no way to disable).

I think it is not enabled inside the CodeAct's system prompt? I remember the delegation browsing PR was not merged yet.

Reply:
Oh, I see you are prompting the agent to do

> for example, for query 'Who is the president of the United States?', you can delegate the task to a browser agent via <execute_browse> Who is the president of the United States? </execute_browse>

That might conflict with the original prompts of CodeActAgent...

Reply:
> I think it is not enabled inside the CodeAct's system prompt? I remember the delegation browsing PR was not merged yet.

It was merged since CodeActAgent 1.6 version, if not 1.5

Reply:
> I think it is not enabled inside the CodeAct's system prompt? I remember the delegation browsing PR was not merged yet.

Ugh nvm, I think we still teach the agent to use `<execute_browse>` to get simple web page stuff.

If this is the case, maybe we can parse the passed along instruction by matching from `f'{thought}. I should start with: {browse_actions}'` regex... it is a bit ugly but i think it is probably fine for eval.

Reply:
I revert it and add some parsing logic in the new commit! @li-boxuan thanks for pointing this out!